src/map/roe.cpp

```diff
@@ -557,4 +601,60 @@ namespace roeutils
         zoneutils::ForEachZone([](CZone* PZone) { PZone->ForEachChar([](CCharEntity* PChar) { ClearWeeklyRecords(PChar); }); });
     }
 
+    // Weekly Ranking Reset
+    void CycleUnityRankings()
+    {
+        const char* rankingQuery = "UPDATE unity_system SET members_prev = members_current, points_prev = points_current, members_current = 0, points_current = 0;";
+        Sql_Query(SqlHandle, rankingQuery);
+
+        roeutils::UpdateUnityRankings();
+    }
+
+    void UpdateUnityRankings()
+    {
+        const char* memberQuery = "UPDATE unity_system JOIN (SELECT unity_leader, COUNT(*) AS members FROM char_profile GROUP BY unity_leader) TMP ON unity_system.leader = unity_leader SET unity_system.members_current = members;";
+        Sql_Query(SqlHandle, memberQuery);
+
+        const char* unityQuery = "SELECT leader, members_prev, points_prev FROM unity_system;";
+        int32       ret        = Sql_Query(SqlHandle, unityQuery);
+
+        std::pair<uint8, int32> unityEval[11];
+
+        if (ret != SQL_ERROR && Sql_NumRows(SqlHandle) != 0)
+        {
+            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
+            {
+                uint8 unityLeader = Sql_GetIntData(SqlHandle, 0);
+
+                if (unityLeader > 0 && unityLeader < 12)
+                {
+                    int32 numMembers = Sql_GetIntData(SqlHandle, 1);
+                    int32 rawEval    = Sql_GetIntData(SqlHandle, 2);
+
+                    unityEval[unityLeader - 1].first  = unityLeader;
+                    unityEval[unityLeader - 1].second = (numMembers > 0) ? rawEval / numMembers : 0;
+                }
+            }
+        }
+
+        std::sort(std::begin(unityEval), std::end(unityEval),
+                  [](std::pair<uint8, uint32> a, std::pair<uint8, uint32> b) { return a.second > b.second; });
+
+        uint8 currentRank = 1;
+        uint8 rankGap     = 1;
+        for (uint8 i = 0; i < 11; i++)
+        {
+            roeutils::RoeSystem.unityLeaderRank[unityEval[i].first - 1] = currentRank;
+
+            if (unityEval[i + 1].second < unityEval[i].second)
+            {
+                currentRank = currentRank + rankGap;
+                rankGap     = 1;
+            }
+            else
+            {
+                rankGap++;
+            }
+        }
```

Comment:
Would we be able to shorten this to something like:
```cpp
        const char* unityQuery = "SELECT leader, members_prev, points_prev FROM unity_system ORDER BY CASE WHEN members_prev = 0 THEN 0 ELSE points_prev/members_prev END DESC;";
        int32       ret        = Sql_Query(SqlHandle, unityQuery);

        if (ret != SQL_ERROR && Sql_NumRows(SqlHandle) != 0)
        {
            uint8 currentRank = 1;
            int32 prev_eval   = 0;
            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
            {
                int32 members  = Sql_GetIntData(SqlHandle, 1);
                int32 points   = Sql_GetIntData(SqlHandle, 2);
                int32 new_eval = members == 0 ? 0 : points / members;
                if (new_eval < prev_eval)
                {
                    currentRank++;
                }
                prev_eval = new_eval;
                roeutils::RoeSystem.unityLeaderRank[Sql_GetIntData(SqlHandle, 0) - 1] = currentRank;
            }
        }
```
It looks like we're doing a lot of extra work when the SQL could sort it for us.

Reply:
Agreed.  I'm an SQL potato, and will update and test from the above suggestions.

Reply:
I believe the above will have a unity with 0 members ranked the same as unities with 0 points but some members. If you want members to count for something can change to this:
```cpp
        const char* unityQuery = "SELECT leader, members_prev, points_prev FROM unity_system ORDER BY CASE WHEN members_prev = 0 THEN -1 ELSE points_prev/members_prev END DESC;";
```
and
```cpp
                int32 new_eval = members == 0 ? -1 : points / members;
```

Reply:
Actually, I think you can even change the query to `SELECT leader, members_prev, points_prev, CASE WHEN members_prev = 0 THEN 0 ELSE points_prev/members_prev END AS eval FROM unity_system ORDER BY eval DESC;` and then use `int32 new_eval = Sql_GetIntData(SqlHandle, 3);`

Reply:
Just getting coffeed up, but I remember why the ranking code is a bit convoluted.  For example, if two unities were tied for first, and three were tied for second, the result would be as follows:

1, 1, 2, 2, 2, 6, 7, ...

Reply:
Aye, the code I posted checks for ties and only increments rank if the evaluation is actually lower. The only place it gets weird is if members + no points is a higher rank than no members + no points, I'm not sure how it's supposed to work, but can be adjusted either way with the `-1` change mentioned.

Reply:
I would assume those should evaluate as the same rank ultimately, since they should be stored with the same rank (0 == 0).  Implementing this change after I finish capturing some more event ids!

Reply:
Also I'm not sure how it's supposed to be handled but you might want to consider float for prev_eval and new_eval and using Sql_GetFloatData.

Reply:
As long as the returned values are floored, I think we're fairly safe.

Reply:
Ah I misunderstood how it was supposed to work. You could add the same "rankGap" system you are currently using into the condensed version (I think this is correct).
```cpp
            uint8 currentRank = 1;
            uint8 rankGap     = 1;
            int32 prev_eval   = 0;
            while (Sql_NextRow(SqlHandle) == SQL_SUCCESS)
            {
                int32 new_eval = Sql_GetIntData(SqlHandle, 1);
                if (new_eval < prev_eval)
                {
                    currentRank = currentRank + rankGap;
                    rankGap     = 1;
                }
                else
                {
                    rankGap++;
                }
                prev_eval = new_eval;
                roeutils::RoeSystem.unityLeaderRank[Sql_GetIntData(SqlHandle, 0) - 1] = currentRank;
            }
```

Reply:
Thanks for that.  Updated the code to reflect your suggestions, though I did change rankGap initialized value to fix an obo that was occurring.

Reply:
Ah nice catch, looks good!